Mathlib/Combinatorics/HalesJewett.lean

```diff
@@ -367,4 +385,90 @@ theorem exists_mono_homothetic_copy {M κ : Type*} [AddCommMonoid M] (S : Finset
     simp_rw [← hy, Option.map_some', Option.getD]
 #align combinatorics.exists_mono_homothetic_copy Combinatorics.exists_mono_homothetic_copy
 
+/-- The type of combinatorial subspaces. A subspace `l : Subspace η α ι` in the hypercube `ι → α`
+defines a function `(η → α) → ι → α` from `η → α` to the hypercube, such that for each coordinate
+`i : ι` and direction `e : η`, the function `fun x ↦ l x i` is either `fun x ↦ x e` for some
+direction `e : η` or constant. We require subspaces to be non-degenerate in the sense that, for
+every `e : η`, `fun x ↦ l x i` is `fun x ↦ x e` for at least one `i`.
+
+Formally, a subspace is represented by a word `l.idxFun : ι → α ⊕ η` which says whether
+`fun x ↦ l x i` is `fun x ↦ x e` (corresponding to `l.idxFun i = Sum.inr e`) or constantly `a`
+(corresponding to `l.idxFun i = Sum.inl a`).
+
+When `α` has size `1` there can be many elements of `Subspace η α ι` defining the same function. -/
+@[ext]
+structure Subspace (η α ι : Type*) where
```

Comment:
Can you redefine `Line` as `Subspace Unit` to avoid duplication?

Reply:
I doubt that's a great idea because `Subspace Unit α ι` coerces to functions `(Unit → α) → ι → α` while `Line α ι` coerces to functions `α → ι → α`. There's not that much duplication to avoid anyway.